dsl/src/main/scala/org/http4s/dsl/impl/Path.scala

```diff
@@ -180,23 +183,44 @@ protected class PathVar[A](cast: String => Try[A]) {
   *      case Root / "user" / IntVar(userId) => ...
   * }}}
   */
-object IntVar extends PathVar(str => Try(str.toInt))
+object IntVar
+    extends PathVar((str: String) =>
+      try
+        Option(str.toInt)
```

Comment:
`Option#apply` makes an additional check that the passed value is not null, which I think is not necessary here, isn't it?

Reply:
Alright, this bikeshedding makes sense. But what would you say about this PR overall? :)

Reply:
From the point of where `PathVar` was before this change, I would say that it makes sense indeed to get rid of the unnecessary `Try` to `Option` conversion.

But honestly I'm not sure I'm really getting the motivation behind the original `PathVar` design out here. I mean not this PR, but all the previous history of its changes. To me this class seems over-designed quite a bit.

First of all, why was it declared as `protected`, what is it for? It is a top-level class afterall, therefore `protected` does not make a lot of sense out there – it just acts as `private [impl]` and nothing more than that.

Moreover, `PathVar` as a whole seems having a shape of a simple old-fashion abstract base class that declares just a single abstract method (i.e. `cast`) and thus its inheritors have to provide their implementations. What was the benefit of making it the non-abstract (but still base) class and requiring to provide an external function instead of just the method implementation? I doubt that it resulted to some better performance somewhere.

My _guess_ (just a guess) is that the original idea was to make its concrete implementations as concise as possible.

I've traced `PathVar` down to its original version and turned out it was as simple as:
https://github.com/http4s/http4s/blob/8845ae00471764f733888109639dc472247ae091/dsl/src/main/scala/org/http4s/dsl/impl/Path.scala#L166-L172
and thus had its descendants declared in a pretty straightforward and concise way:
```scala
object IntVar extends NumericPathVar(_.toInt)
object LongVar extends NumericPathVar(_.toLong)
```

To me it looks like it should be possible to combine all these approaches (including the optimization you made by dropping `Try` out) into something like this:
```scala
class PathVar[A](cast: String => A) {
  def unapply(str: String): Option[A] =
    if (str.nonEmpty)
      try
        Some(cast(str))
      catch {
        case t if scala.util.control.NonFatal(t) => None
      }
    else
      None
}
```
I.e. we could get rid of the `Try => Option` conversion completely while keeping the concrete implementations as concise as it was initially supposed to (if my guess is correct, of course), therefore avoiding any repetitions in there:
```scala
object IntVar extends PathVar(_.toInt)
object LongVar extends PathVar(_.toLong)
object UUIDVar extends PathVar(java.util.UUID.fromString)
```

Wdyt?

Reply:
Oh my Gosh, thanks for all of your ideas on this, really a lot of things to think about. I wasn't willing to rethink/recreate this whole class initially. It feels like worth another outer of this discussion. Isn't it?

> I.e. we could get rid of the Try => Option conversion completely while keeping the concrete implementations as concise as it was initially supposed to (if my guess is correct, of course), therefore avoiding any repetitions in there:

IIUC, you're suggesting using the total function here. Well, I'm not sure it's a good idea. I can easily imagine that users might want to use additional data verification. Like, parse non-negative ints only. Perhaps partial functions fit better for that. That said, exposing an `Option`-based API should be fine too.

Reply:
> I can easily imagine that users might want to use additional data validation. 

Good point, agree. So yeah, it also could be 
```
class PathVar[A](cast: String => Option[A]) {
  def unapply(str: String): Option[A] =
    if (str.nonEmpty)
      try cast(str)
      catch { case t if scala.util.control.NonFatal(t) => None }
    else
      None
}
```
or, as you suggested, `PartialFunction[String, A]`

This way it becomes almost identical to what you already have in the PR, with the only difference that `try...catch` is generalized in the base class rather than copied among `PathVar` inheritors.

Basically in my prospective, `PathVar` with the `cast` function as a class parameter makes only sense if those `cast` functions passed by inheritors are really short and concise. Something like `{ _.toInt }` or at least `{ s => Some(s.toInt) }`. Otherwise IMO it is not worth it and would be better simply to make it an abstract base class like this:
```scala
abstract class PathVar[A] {
  def cast(s: String): Option[A] // abstract method
  def unapply(): Option[A] = ... // implementation as above
}